### DISCUSSION TITLE
GitHub Actions: Filter on release

### DISCUSSION THREAD

**Discussion by sidvishnoi** [community/community#27053] [Actions]
<p>Currently, on:release doesn’t support any filters (for example, on:pull_request supports branches filter). Would be nice if something like following gets supported:</p>
<pre><code>on:
  release:
    events:
    - published
</code></pre>
<p>As a workaround, I’ve to do the following, which leads to one failed build and one successful build on every published release:</p>
<pre><code>on: release
jobs:
  publish:
    steps:
      - name: Cancel if not a release publish # workaround has `on` doesn't have this filter
        run: exit 1
        if: github.event.action != 'published'
      - uses: actions/checkout@v1
</code></pre>

**Comment by cvergne** [comment]
<p>To be relevant with webhooks (as it refers to in actions documentation), I would suggest :</p>
<pre><code>on:
  release:
    actions:
    - published
</code></pre>
<p>Because “published”, etc. are refered in an “action” node in webhooks payload.</p>
<p>But I totally agree with the suggestion because currently it triggers 2 runs for each release.</p>
<p>(Btw, thanks for the workaround)</p>

**Comment by nyahbhinghiPrincess** [comment]
<p>Hi <a class="mention" href="/sidvishnoi">@sidvishnoi</a>,</p>
<p>Thanks for this feedback! We’re always working to improve GitHub, and we consider every suggestion we receive. I’ve logged your feature request in our internal feature request list. Though I can’t guarantee anything or share a timeline for this, I can tell you that it’s been shared with the appropriate teams for consideration.</p>

**Comment by badsyntax** [comment, marked as answer]
<p>You can now filter the release event using types, for example:</p>
<pre><code>on:
  release:
    types: [published]
</code></pre>
<p>More info: <a href="https://help.github.com/en/articles/workflow-syntax-for-github-actions#onevent_nametypes" rel="nofollow noopener">https://help.github.com/en/articles/workflow-syntax-for-github-actions#onevent_nametypes</a></p>

**Comment by sidvishnoi** [comment]
<p>Awesome! Can’t wait to try it out!</p>

**Comment by riker09** [comment]
<p>Are you planning on supporting additional filters?</p>
<p>My codebase is a large monorepo. Being able to filter based on release tag would be a huge plus.</p>
<p>Here’s what I’m thinking about:</p>
<pre><code>on:
  release:
    types: [published, created, edited]
    tags:
      - service-a
      - service-b
      - !service-xyz
      - some-other-stuff-v2.*
</code></pre>
<p>I hope you get the idea.</p>

**Comment by badsyntax** [comment]
<p>I guess you can use the step conditionals to filter by tag value? Not sure. Sorry I’m not a GitHub support engineer.</p>
<p>Perhaps create a new thread?</p>

**Comment by riker09** [comment]
<p>Filtering on the actual job and/or step is exactly what I’m doing right now. However, since I have a monorepo there are quite a few workflows that get triggered on the “release” event only to be discarded a few moments later.</p>
<p>Here’s an example of the workflow I’m aiming for: Say I have two things (services, packages, …) in my repository that I want to build a package of (Docker, NPM, …), “thing-a” and “thing-b” living in their respective subfolders. When I create a release and tag it with “thing-a@v1.3.9” I want to trigger the build &amp; publish workflow of “thing-a”. I can do this with an “if” filter at the job:</p>
<pre><code>on:
release:
type: [published]

jobs:
  build_and_publish:
    runs-on: ubuntu-latest
    if: github.event_name == 'release' &amp;&amp; contains(github.ref, '/thing-a')
    steps:
      - ...
</code></pre>
<p>There is an identical workflow for “thing-b”. Now when I publish a new release both workflows trigger. Workflow “thing-b” is then stopped at  “job.build_and_publish.if” however the triggered workflow appears in the list.</p>
<p>With the new filter system at the workflow overview in place I admit that this is rather an annoyance than a real issue, but since there are similar filters available for the “push” workflow trigger I thought this would make a great addition. And you are right, a new topic would probably make more sense. :slight_smile:</p>

**Comment by mpdude** [comment]
<p>Does anybody know what exactly the difference is between the published, unpublished, created, edited, deleted or prereleased types?</p>
<p>The documentation does not details this.</p>

**Comment by sagikazarmark** [comment]
<p>It’s not easy to find, but it’s documented here: <a href="https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/webhook-events-and-payloads#release" rel="noopener nofollow ugc">https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/webhook-events-and-payloads#release</a></p>

**Comment by AnderssonPeter** [comment]
<aside class="quote no-group" data-username="riker09" data-post="6" data-topic="16058">
<div class="title">
<div class="quote-controls"></div>
 riker09:</div>
<blockquote>
<pre><code class="lang-auto">on:
  release:
    types: [published, created, edited]
    tags:
      - service-a
      - service-b
      - !service-xyz
      - some-other-stuff-v2.*
</code></pre>
</blockquote>
</aside>
<p>This syntax would be awesome for monorepos!<br>
Is there any news if this is going to be implemented? and if not could someway to cancel a action without yielding a error?</p>
<p>I found a way to do this by using the <code>if:</code> on the job.</p>
<pre><code class="lang-auto"># This is a basic workflow to help you get started with Actions

name: CI

# Controls when the action will run. 
on:
  release:
    types: [created]

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "build"
  build-api:
    if: startsWith(github.ref, 'refs/tags/api_v.')
</code></pre>

**Comment by doflo-dfa** [comment]
<p>This is the solution we have however, it is not as good as allowing some type of filter in the on: clause as it creates builds in the actions list that start and immediately complete as opposed to not listed like actions that are skipped due to the on clause</p>
